torch/_tensor.py

```diff
@@ -383,16 +409,26 @@ def _reduce_ex_internal(self, proto):
             type(self) is not torch.Tensor
             and type(self).__torch_dispatch__ is not torch.Tensor.__torch_dispatch__
             and (
-                isinstance(
-                    self,
-                    (
-                        torch._subclasses.fake_tensor.FakeTensor,
-                        torch._subclasses.functional_tensor.FunctionalTensor,
-                    ),
+                (
+                    isinstance(
+                        self, torch._subclasses.functional_tensor.FunctionalTensor
+                    )
+                    or (
+                        isinstance(self, torch._subclasses.fake_tensor.FakeTensor)
+                        and not metadata_only
+                    )
+                )
+                or (
+                    not isinstance(
+                        self,
+                        (
+                            torch._subclasses.functional_tensor.FunctionalTensor,
+                            torch._subclasses.fake_tensor.FakeTensor,
+                        ),
+                    )
+                    and self.data_ptr() == 0
```

Comment:
Before the `or self.data_ptr() == 0` would cause `self.data_ptr()` to always be called.

We shouldn't be calling `data_ptr` for FakeTensor that yields a deprecation message, this fixes that

```
<stdin>:1: UserWarning: Accessing the data pointer of FakeTensor is deprecated and will error in PyTorch 2.5. This is almost definitely a bug in your code and will cause undefined behavior with subsystems like torch.compile. Please wrap calls to tensor.data_ptr() in an opaque custom op; If all else fails, you can guard accesses to tensor.data_ptr() on isinstance(tensor, FakeTensor). (Triggered internally at /data/users/mg1998/pytorch/c10/core/StorageImpl.cpp:34.)
```

Reply:
Wouldn't the "or" statement there make it not called for fake or functional Tensors?

Reply:
You're right, I misread the code

Reply:
edit: Ah I see, I added the new bit because my new condition would make it go into the or statement, my bad